### PR TITLE
Add MMEarth dataset

### PR DESCRIPTION
**Pull request by nilsleh**
This PR adds the [MMEarth](https://vishalned.github.io/mmearth/) dataset. Implementation is tested with the 100k version under the assumption that the dataset format is identical for the other versions.

There are some expected [changes to the dataset](https://github.com/vishalned/MMEarth-train/issues/5) so still a draft. But opening a PR already for some discussion points:

- As a multi modal dataset there are various configuration and selection schemes. The dataset has arguments for modality and band selection within modalities (the current implementation is a suggestion, but open for cleaner suggestions)
- For that same reason, we don't have a classic "image" and "label" sample dictionary because there are various ways of using this dataset across classification, segmentation and pixelwise regression. Is it reasonable to expect people to write `transform` functions that would define the sample as needed for their models?
- Normalization statistics are provided with the dataset. Normally we do standardization in the DataModule, however, given the complexity of various modalities and also two normalization schemes (z-score and min-max) I think it makes the dataset more useful when supplying this functionality in the dataset already
- The most complex part is handling the different modalities because some of them require some extra handling as seen [here](https://github.com/vishalned/MMEarth-train/blob/8d6114e8e3ccb5ca5d98858e742dac24350b64fd/mmearth_dataset.py#L69)
- Additionally, need to decide what to do with the nan-values and clearly comment that
- currently this is just an implementation to support a non-geo dataset with standard index sampling, for other indexing schemes we can write an inherited class that also has a script that generates the needed parquet file and download it from HF
- probably things I am missing so feel free to add @adamjstewart @ando-shah

### CODE REVIEW

**Comment by nilsleh** [comment]
@adamjstewart and @ando-shah I think as a basic Dataset this does the job. For any more specific requirements we can subclass this to support other model specific needs.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:75]: Use tuples instead of lists to make ruff happy

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:77]:
```suggestion
    filenames: ClassVar[dict[str, str]] = {
```

**Review by adamjstewart — COMMENTED** [review]
I added a new test to ruff

**Comment by adamjstewart** [comment]
Planning to get a working draft of SatlasPretrain first and then we can modify this PR to ensure it matches and meets @ando-shah's needs. These will be some of our first multi-modal datasets, so want to make sure we're doing things consistently (at least at an external user-facing API level).

File: docs/api/non_geo_datasets.csv
Comment (adamjstewart) [inline docs/api/non_geo_datasets.csv:27]:
```suggestion
`MMEarth`_,"C, S", Google Earth,"CC-BY-4.0","100K--1M",,"128x128 or 64x64","multiple modalities",MSI
```
This will create an en-dash for ranges.

File: tests/datasets/test_mmearth.py
Comment (adamjstewart) [inline tests/datasets/test_mmearth.py:20]: Do these need os.path.join so it uses \ on Windows? Or does `Path` handle that internally?

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:21]:
```suggestion
    """MMEarth dataset.
```
Same elsewhere, this is the official dataset name

File: torchgeo/datamodules/mmearth.py
Comment (adamjstewart) [inline torchgeo/datamodules/mmearth.py:39]: What is the purpose of this override? If you don't want any augs, that should be much easier to override.

File: torchgeo/datamodules/mmearth.py
Comment (adamjstewart) [inline torchgeo/datamodules/mmearth.py:1]: I'm dubious of the value of a data module that isn't compatible with any of our trainers, but it also isn't hurting anything.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:28]:
```suggestion
    * MMEarth: 128x128 px, 1.2M tiles, 579 GB
    * MMEarth64: 64x64 px, 1.2M tiles, 162 GB
    * MMEarth100k: 128x128 px, 100K tiles, 48 GB
```
these aren't units

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:37]:
```suggestion
    * ERA5: Climate reanalysis data for temperature mean, min, and max of [year, month, previous month]
```
Since everything else is capitalized

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:56]:
```suggestion
    * JSON files for band statistics, splits, and tile information
```

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:60]: If we have time, let's add download support

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:75]: Let's use tuples instead of lists

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:84]: A normal tuple is fine here. Function inputs should be as broad as possible. Function outputs (and class attributes) should be as narrow as possible.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:189]: How much added complexity was it to support all 3? If it wasn't hard, great! If it's annoying, we could stick to the big version only.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:178]: No need to prefix dataset parameters with `ds_`. I'm a little worried about calling this version because there is also a 001 version that may change as well. Not sure what a good alternative would be though.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:241]: Not true. I usually move the errors to the `__init__` docstring since `_verify` won't be included in the docs.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:288]: Missing documentation of return value

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:504]: Missing documentation of return value

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:314]: Validation is good, but not if it makes the code too complex.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: torchgeo/datasets/mmearth.py
Comment (nilsleh) [reply torchgeo/datasets/mmearth.py:370]: Okay, I think we can provide a guess on a default, but for some modalities like elevation or vegetation height, they might be used as either "image" or "mask".

File: torchgeo/datasets/mmearth.py
Comment (nilsleh) [reply torchgeo/datasets/mmearth.py:191]: How about "variant", or "subset"

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [reply torchgeo/datasets/mmearth.py]: We just need a good default. Someone can customize/override/transform the dataset if they have a different use case. Check other datasets and see what defaults they used.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [reply torchgeo/datasets/mmearth.py:191]: I like `subset`

**Comment by nilsleh** [comment]
Ando requested that Sentinel 1 should be separated into ascending and descending. That turned out to be a bit tricky, since MMEarth only stores "sentinel1" as a concatenation of ascending and descending, but I think I implemented it correctly to support band selection and normalization etc

**Comment by nilsleh** [comment]
As another point regarding Sentinel1:

- Sentinel 1 ascending and descending are stored separately, however, in the hdf5 file, there is just a single sentinel1 key that always has a shape of [8, 128, 128] so if one of ascending or descending or a subset of bands within the two is not available it is still stored as [1, 128,128] of nans in the dataset
- in order to not return Nans from the `__getitem__` , only non-nan data is now returned, i.e. if I request `{'sentinel1_asc': ['VV', 'VH', 'HH', 'HV']}` in the dataset init but for the particular tile only `{'sentinel1_asc': ['VV', 'VH']}` is available, then a [2, 128, 128] tensor will be returned
- additionally, the sample contains a dict with the sensor and bands actually found in this particular sample

File: docs/api/datasets/non_geo_datasets.csv
Comment (adamjstewart) [inline docs/api/datasets/non_geo_datasets.csv:31]: Can we list out the resolutions instead of just saying "multiple modalities"?

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:552]: I think this should be a datamodule transform, not hard-coded into the dataset.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:75]: No need to type hint here, the type can be automatically detected

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:83]: Same here

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:493]:
```suggestion
            # first replace 0 with nan then assign new labels to have 0-index classes
```

**Review by adamjstewart — COMMENTED** [review]
Data loader seems much more complicated than it needs to be. Does it satisfy all of Ando's requirements?

**Comment by nilsleh** [comment]
> Data loader seems much more complicated than it needs to be. Does it satisfy all of Ando's requirements?

Yeah all the feedback I have gotten from him, I have incorporated. I agree it is complicated but it's also loading data from >10 modalities so not a typical single image/target dataset. He said he will try it now and give any other feedback.

File: torchgeo/datasets/mmearth.py
Comment (nilsleh) [reply torchgeo/datasets/mmearth.py:551]: I think I would disagree with you here. I know that for most datasets, the normalization in torchgeo is only applied in the datamodule. I am also not a big fan of that (I think normalization and augmentation should be handled separately). Thinking ahead of training over multiple sources, normalization is unique to each data source, while things like geometric augmentations (flips etc) are applicable to all inputs. A datamodule then would have to involve all this normalization logic but for multiple cases depending on the data source, so to me this makes more sense.

File: docs/api/datasets/non_geo_datasets.csv
Comment (nilsleh) [reply docs/api/datasets/non_geo_datasets.csv:31]: The paper on page 6 states ". Additionally, all pixel-level modalities were
re-projected to the Sentinel-2 10m grid, if needed, yielding a harmonized data
cube." So I guess actually everything is 10m in this dataset.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [reply torchgeo/datasets/mmearth.py:551]: My counterarguments:

* TorchGeo is based on torchvision, which does not do normalization in the dataset, although you can specify transforms you want to apply
* Lightning intends its datamodules to be used for [splitting, transforming, normalizing, and tokenizing](https://lightning.ai/docs/pytorch/stable/data/datamodule.html#why-do-i-need-a-datamodule) the data
* Yes, normalizations are dataset-specific, but so are datamodules. We don't need generic datamodules that can be used for any dataset

I agree that multimodal datasets throw a wrench in this design, since each image needs a different normalization. I don't have a good counterargument for that, it is kind of annoying that we have to write custom transforms to handle each mode separately. I also fully agree that normalizations should be parameters so that I can specify them in my YAML file, although Lightning doesn't have a way to checkpoint these.

I don't disagree with you enough to hold up this PR, but I would like to decide on a consistent strategy for all datasets. Right now, our datasets are hundreds of lines of code, while our datamodules are only tens of lines of code. I want to try to shift as much of the complexity as possible to the datamodules if we can (for example, converting mask classes to ordinal numbers) so that the data loaders can focus on data loading and all transforms can be applied on the GPU.

File: docs/api/datasets/non_geo_datasets.csv
Comment (adamjstewart) [reply docs/api/datasets/non_geo_datasets.csv]: Maybe @vishalned can comment (FYI, we are adding a data loader for MMEarth to TorchGeo!)

Comment (vishalned) [reply docs/api/datasets/non_geo_datasets.csv]: Hey, yes exactly. every modality has been reprojected to 10m resolution. 

Also, just reading from the previous comments it seems like its a bit tricky with the nan values and the S1 bands (sorry about this). Incase the hdf5 is not useful to detect empty bands, i believe i also store some meta information regarding the bands (which are present and not) in the tile_info.json. 

Hope this helps, and let me know if you need any more clarifications

File: torchgeo/datasets/mmearth.py
Comment (nilsleh) [reply torchgeo/datasets/mmearth.py:551]: Yeah I am happy to discuss this at a later stage, just as a brief answer:

- torchvision expects the normalization on the dataset level. if you take any common dataset like imagenet, mnist, cifar etc,  everyone always has to copy that same transform snippet and normalization values from some other repo and pass it to the dataset class, which is redudant and seems unnecessary
- in the documentation lightning page in their example, the normalization is defined in a `DataModule` but is actually applied inside the `Dataset` through a transform function, no `after_batch_transform` function

Thus I don't think even lightning is fully clear on this, but at the top of my head I also don't have the perfect solution for torchgeo.

Comment (ando-shah) [reply torchgeo/datasets/mmearth.py:551]: Im happy to move the normalization function to wrappers to keep this streamlined

Comment (nilsleh) [reply torchgeo/datasets/mmearth.py:551]: @adamjstewart would you prefer, if I write a datamodule where all the normalization logic lives?

Comment (adamjstewart) [reply torchgeo/datasets/mmearth.py:551]: I'll let you decide. I'm okay with both. I'm just trying to find ways to keep the datasets simple.

Comment (nilsleh) [reply torchgeo/datasets/mmearth.py:551]: Fair! I think I would keep it for now then, but open to changing it at some other point.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:198]: At the moment, `all_modality_bands` isn't documented. Should it be?

**Comment by nilsleh** [comment]
@ando-shah found a bug in my implementation. The available band names were different from the sample specific band name, for example in era5, they have the current or previous date of the sample included in the band name, so not a general name, and therefore it was not correctly retrieved.

File: torchgeo/datasets/mmearth.py
Comment (adamjstewart) [inline torchgeo/datasets/mmearth.py:204]:
```suggestion
            AssertionError: if *normalization_mode*, *subset*, or *split* is invalid
```
The Python and Sphinx docs use italics to reference parameters, trying to match them.

File: torchgeo/datasets/mmearth.py
Comment (nilsleh) [reply torchgeo/datasets/mmearth.py]: I meant to point to the class variable `all_modality_bands`

**Comment by nilsleh** [comment]
@adamjstewart @ando-shah while trying to write a subclass based on a metadata file, an annoying part is the `image_{}` and `mask_{}` naming for the sample keys, because you need to remove them again to match the meta data naming of the modalities. So since there isn't a torchgeo trainer that works out of the box on this dataset anyways, maybe we can just return the modality name itself?

**Comment by nilsleh** [comment]
@admjstewart okay I think we can merge this then, and handle everything else in a subclass.

File: torchgeo/datasets/mmearth.py
Comment (nilsleh) [reply torchgeo/datasets/mmearth.py]: mhm, I thought I could have the reference to the existing class attributes, do I need to redefine them again?

Comment (adamjstewart) [reply torchgeo/datasets/mmearth.py]: The class attributes exist in the code, but not in the documentation. To make them exist in the documentation too, you need to use `#:` to describe the attribute. You don't need to define them twice.

Comment (nilsleh) [reply torchgeo/datasets/mmearth.py]: Mhm, but it's all displayed in uppercase on the docs page, which will be misleading because that is not how the actual data is defined. Is there a way to control that?

**Review by adamjstewart — APPROVED** [review]
Still a bit more complicated than I would like, but it gets the job done.

**Comment by adamjstewart** [comment]
Just realized we forgot to add a plot method to this dataset: #2377 

I'll see if any of the remote sensing seminar students are interested in adding this. If not, may ask you to add a plot after deadlines pass and you have more time.

**Pull request merged**
